src/hardware/input/intel8042.cpp

```diff
@@ -0,0 +1,1044 @@
+		}
+	}
+}
+
+static void buffer_add(const uint8_t byte,
+                       const bool is_from_aux = false,
+                       const bool is_from_kbd = false,
+                       const bool skip_delay  = false)
+{
+	if ((is_from_aux && is_disabled_aux) || (is_from_kbd && is_disabled_kbd)) {
+		// Byte came from a device which is currently disabled
+		return;
+	}
+
+	if (buffer_num_used >= buffer_size) {
+		warn_buffer_full();
+		flush_buffer();
+		return;
+	}
+
+	const size_t idx = (buffer_start_idx + buffer_num_used++) % buffer_size;
+
+	if (is_from_kbd && uses_kbd_translation) {
+		buffer[idx].data = get_translated(byte);
+	} else {
+		buffer[idx].data = byte;
+	}
+	buffer[idx].is_from_aux = is_from_aux;
+	buffer[idx].is_from_kbd = is_from_kbd;
+	buffer[idx].skip_delay  = skip_delay || (!is_from_aux && !is_from_kbd);
+
+	if (is_from_aux) {
+		++waiting_bytes_from_aux;
+	}
+	if (is_from_kbd) {
+		++waiting_bytes_from_kbd;
+	}
+
+	maybe_transfer_buffer();
+}
+
+static void buffer_add_aux(const uint8_t byte, const bool skip_delay = false)
+{
+	const bool is_from_aux = true;
+	const bool is_from_kbd = false;
+
+	buffer_add(byte, is_from_aux, is_from_kbd, skip_delay);
+}
+
+static void buffer_add_kbd(const uint8_t byte)
+{
+	const bool is_from_aux = false;
+	const bool is_from_kbd = true;
+
+	buffer_add(byte, is_from_aux, is_from_kbd);
+}
+
+// ***************************************************************************
+// Command handlers
+// ***************************************************************************
+
+static uint8_t get_input_port() // aka port P1
+{
+	static union {
+		uint8_t data = 0b1010'0000;
+
+		// bit 0: keyboard data in, ISA - unused
+		// bit 1: mouse data in, ISA - unused
+		// bit 2: ISA, EISA, PS/2 - unused
+		//        MCA - 0 = keyboard has power, 1 = no power
+		//        might be configured for clock switching
+		// bit 3: ISA, EISA, PS/2 - unused
+		//        might be configured for clock switching
+
+		// bit 4: 0 = 512 KB, 1 = 256 KB
+		// bit 5: 0 = manufacturer jumper, infinite diagnostics loop
+		// bit 6: 0 = CGA, 1 = MDA
+		bit_view<6, 1> lacks_cga;
+		// bit 7: 0 = keyboard locked, 1 = not locked
+
+	} port;
+
+	port.lacks_cga = (machine < MCH_CGA);
+
+	return port.data;
+}
+
+static uint8_t get_output_port() // aka port P2
+{
+	static union {
+		uint8_t data = 0b0000'0001;
+
+		// bit 0: 0 = CPU reset, 1 = normal
+		// bit 1: 0 = A20 disabled, 1 = enabled
+		bit_view<1, 1> is_a20_enabled;
+		// bit 2: mouse data out, ISA - unused
+		// bit 3: mouse clock, ISA - unused
+
+		// bit 4: 0 = IRQ1 (keyboard) not active, 1 = active
+		bit_view<4, 1> is_irq_active_kbd;
+		// bit 5: 0 = IRQ12 (mouse) not active, 1 = active
+		bit_view<5, 1> is_irq_active_aux;
+		// bit 6: keyboard clock
+		// bit 7: keyboard data out
+
+	} port;
+
+	port.is_a20_enabled    = MEM_A20_Enabled();
+	port.is_irq_active_kbd = is_irq_active_kbd ? 1 : 0;
+	port.is_irq_active_aux = is_irq_active_aux ? 1 : 0;
+
```

Comment:
(just mentioning it - single-bit `bit_view`s can be set directly using a bool; there's explicit 1-bit handling for `bool` types because it's unambiguous)

Reply:
Does not work for me (GCC 12.2.0)

```
In file included from ../src/hardware/input/intel8042.cpp:25:
../include/bit_view.h: In instantiation of 'constexpr void bit_view<view_index, view_width>::check_rhs(rhs_type) [with rhs_type = bit_view<0, 1>; int view_index = 4; int view_width = 1]':
../include/bit_view.h:183:12:   required from 'constexpr bit_view<view_index, view_width>& bit_view<view_index, view_width>::operator=(rhs_type) [with rhs_type = bit_view<0, 1>; int view_index = 4; int view_width = 1]'
../src/hardware/input/intel8042.cpp:558:27:   required from here
../include/bit_view.h:131:59: error: static assertion failed: the bit_view's value can only accept integral types
  131 |                 static_assert(std::is_integral<rhs_type>::value,
      |                                                           ^~~~~
../include/bit_view.h:131:59: note: 'std::integral_constant<bool, false>::value' evaluates to false
../include/bit_view.h: In instantiation of 'constexpr void bit_view<view_index, view_width>::check_rhs(rhs_type) [with rhs_type = bit_view<1, 1>; int view_index = 5; int view_width = 1]':
../include/bit_view.h:183:12:   required from 'constexpr bit_view<view_index, view_width>& bit_view<view_index, view_width>::operator=(rhs_type) [with rhs_type = bit_view<1, 1>; int view_index = 5; int view_width = 1]'
../src/hardware/input/intel8042.cpp:559:27:   required from here
../include/bit_view.h:131:59: error: static assertion failed: the bit_view's value can only accept integral types
../include/bit_view.h:131:59: note: 'std::integral_constant<bool, false>::value' evaluates to false
```

Reply:
Fixing

Reply:
Thanks for the patience with this, @FeralChild64.
`main` should now be able to handle assignment:

```c++
port.is_irq_active_kbd = is_irq_active_kbd;
port.is_irq_active_aux = is_irq_active_aux;
```

(also has new static assert to ensure the LHS bit-view has enough bits to accommodate the RHS bit-view).

Reply:
Nice one @kcgen!

Reply:
> (just mentioning it - single-bit bit_views can be set directly using a bool; there's explicit 1-bit handling for bool types because it's unambiguous)

The reason this didn't work is that the right-hand-side values are actually single-bit `bit_view` aliases with a slightly different layout than the LHS (which should all just "work", but didn't in the prior bit_view code). Very nice use of all this, @FeralChild64!

Reply:
OK, now this works - changed.